Handle GCS bucket 503s

Ref #810 

Some deployments are failing due to failed bucket uploads.  This is one of our most common deployment errors.  The source appears to be a 503 from GCP:

    I 2020-12-03T10:51:28.599964831Z [ERROR] dshop.logic.deploy: Unknown error deploying to bucket.
    I 2020-12-03T10:51:28.630321577Z [ERROR] dshop.logic.deploy: ApiError: Backend Error
        at new ApiError (/app/node_modules/@google-cloud/common/build/src/util.js:59:15)
        at Util.parseHttpRespBody (/app/node_modules/@google-cloud/common/build/src/util.js:194:38)
        at Util.handleResp (/app/node_modules/@google-cloud/common/build/src/util.js:135:117)
        at /app/node_modules/@google-cloud/common/build/src/util.js:434:22
        at onResponse (/app/node_modules/retry-request/index.js:214:7)
        at /app/node_modules/teeny-request/build/src/index.js:150:17
        at runMicrotasks (<anonymous>)
        at processTicksAndRejections (internal/process/task_queues.js:93:5) {
    I 2020-12-03T10:51:28.630408176Z   code: 503,
    I 2020-12-03T10:51:28.630412847Z   errors: [
    I 2020-12-03T10:51:28.630417330Z     {
    I 2020-12-03T10:51:28.630422183Z       message: 'Backend Error',
    I 2020-12-03T10:51:28.630426768Z       domain: 'global',
    I 2020-12-03T10:51:28.630431399Z       reason: 'backendError'
    I 2020-12-03T10:51:28.630436005Z     }
    I 2020-12-03T10:51:28.630440726Z   ],
    I 2020-12-03T10:51:28.630445105Z   response: PassThrough {
    I 2020-12-03T10:51:28.630449050Z     _readableState: ReadableState {
    I 2020-12-03T10:51:28.630452240Z       objectMode: false,
    I 2020-12-03T10:51:28.630455421Z       highWaterMark: 16384,
    I 2020-12-03T10:51:28.630458490Z       buffer: BufferList { head: null, tail: null, length: 0 },
    I 2020-12-03T10:51:28.630462911Z       length: 0,
    I 2020-12-03T10:51:28.630467320Z       pipes: [],
    I 2020-12-03T10:51:28.630472051Z       flowing: true,
    I 2020-12-03T10:51:28.630476370Z       ended: true,
    I 2020-12-03T10:51:28.630480838Z       endEmitted: true,
    I 2020-12-03T10:51:28.630485497Z       reading: false,
    I 2020-12-03T10:51:28.630490251Z       sync: false,
    I 2020-12-03T10:51:28.630494871Z       needReadable: false,
    I 2020-12-03T10:51:28.630499521Z       emittedReadable: false,
    I 2020-12-03T10:51:28.630504056Z       readableListening: false,
    I 2020-12-03T10:51:28.630508667Z       resumeScheduled: false,
    I 2020-12-03T10:51:28.630513242Z       errorEmitted: false,
    I 2020-12-03T10:51:28.630517851Z       emitClose: true,
    I 2020-12-03T10:51:28.630522414Z       autoDestroy: true,
    I 2020-12-03T10:51:28.630527117Z       destroyed: true,
    I 2020-12-03T10:51:28.630531564Z       errored: null,
    I 2020-12-03T10:51:28.630534835Z       closed: true,
    I 2020-12-03T10:51:28.630537860Z       closeEmitted: true,
    I 2020-12-03T10:51:28.630540960Z       defaultEncoding: 'utf8',
    I 2020-12-03T10:51:28.630544176Z       awaitDrainWriters: null,
    I 2020-12-03T10:51:28.630547248Z       multiAwaitDrain: false,
    I 2020-12-03T10:51:28.630550687Z       readingMore: false,
    I 2020-12-03T10:51:28.630553741Z       decoder: null,
    I 2020-12-03T10:51:28.630556788Z       encoding: null,
    I 2020-12-03T10:51:28.630559904Z       [Symbol(kPaused)]: false
    I 2020-12-03T10:51:28.630562975Z     },
    I 2020-12-03T10:51:28.630566104Z     _events: [Object: null prototype] {
    I 2020-12-03T10:51:28.630573894Z       prefinish: [Function: prefinish],
    I 2020-12-03T10:51:28.630577113Z       error: [Array],
    I 2020-12-03T10:51:28.630580146Z       data: [Function (anonymous)],
    I 2020-12-03T10:51:28.630583239Z       end: [Function (anonymous)]
    I 2020-12-03T10:51:28.630586310Z     },
    I 2020-12-03T10:51:28.630589364Z     _eventsCount: 4,
    I 2020-12-03T10:51:28.630592404Z     _maxListeners: undefined,
    I 2020-12-03T10:51:28.630595490Z     _writableState: WritableState {
    I 2020-12-03T10:51:28.630598577Z       objectMode: false,
    I 2020-12-03T10:51:28.630601631Z       highWaterMark: 16384,
    I 2020-12-03T10:51:28.630604716Z       finalCalled: false,
    I 2020-12-03T10:51:28.630607768Z       needDrain: false,
    I 2020-12-03T10:51:28.630610808Z       ending: true,
    I 2020-12-03T10:51:28.630613821Z       ended: true,
    I 2020-12-03T10:51:28.630616838Z       finished: true,
    I 2020-12-03T10:51:28.630619879Z       destroyed: true,
    I 2020-12-03T10:51:28.630622964Z       decodeStrings: true,
    I 2020-12-03T10:51:28.630627634Z       defaultEncoding: 'utf8',
    I 2020-12-03T10:51:28.630630864Z       length: 0,
    I 2020-12-03T10:51:28.630634016Z       writing: false,
    I 2020-12-03T10:51:28.630637062Z       corked: 0,
    I 2020-12-03T10:51:28.630640107Z       sync: false,
    I 2020-12-03T10:51:28.630643138Z       bufferProcessing: false,
    I 2020-12-03T10:51:28.630646200Z       onwrite: [Function: bound onwrite],
    I 2020-12-03T10:51:28.630649417Z       writecb: null,
    I 2020-12-03T10:51:28.630652454Z       writelen: 0,
    I 2020-12-03T10:51:28.630655474Z       afterWriteTickInfo: null,
    I 2020-12-03T10:51:28.630658534Z       buffered: [],
    I 2020-12-03T10:51:28.630661551Z       bufferedIndex: 0,
    I 2020-12-03T10:51:28.630664583Z       allBuffers: true,
    I 2020-12-03T10:51:28.630667596Z       allNoop: true,
    I 2020-12-03T10:51:28.630670643Z       pendingcb: 0,
    I 2020-12-03T10:51:28.630673662Z       prefinished: true,
    I 2020-12-03T10:51:28.630676687Z       errorEmitted: false,
    I 2020-12-03T10:51:28.630679747Z       emitClose: true,
    I 2020-12-03T10:51:28.630682764Z       autoDestroy: true,
    I 2020-12-03T10:51:28.630685808Z       errored: null,
    I 2020-12-03T10:51:28.630688848Z       closed: true
    I 2020-12-03T10:51:28.630691868Z     },
    I 2020-12-03T10:51:28.630694873Z     allowHalfOpen: true,
    I 2020-12-03T10:51:28.630697884Z     statusCode: 503,
    I 2020-12-03T10:51:28.630700921Z     statusMessage: 'Service Unavailable',
    I 2020-12-03T10:51:28.630704005Z     request: {
    I 2020-12-03T10:51:28.630710515Z       agent: [Agent],
    I 2020-12-03T10:51:28.630713769Z       headers: [Object],
    I 2020-12-03T10:51:28.630716968Z       href: 'https://storage.googleapis.com/upload/storage/v1/b/ds-deploy-mainnet-ultimate-franchise-fantasy-sports/o?uploadType=multipart&name=images%2Fnew-shop%2Fphoto.svg&predefinedAcl=publicRead'
    I 2020-12-03T10:51:28.630720781Z     },
    I 2020-12-03T10:51:28.630723771Z     body: '{\n' +
    I 2020-12-03T10:51:28.630726933Z       '  "error": {\n' +
    I 2020-12-03T10:51:28.630730199Z       '    "code": 503,\n' +
    I 2020-12-03T10:51:28.630733571Z       '    "message": "Backend Error",\n' +
    I 2020-12-03T10:51:28.630736980Z       '    "errors": [\n' +
    I 2020-12-03T10:51:28.630740159Z       '      {\n' +
    I 2020-12-03T10:51:28.630743297Z       '        "message": "Backend Error",\n' +
    I 2020-12-03T10:51:28.630748540Z       '        "domain": "global",\n' +
    I 2020-12-03T10:51:28.630751873Z       '        "reason": "backendError"\n' +
    I 2020-12-03T10:51:28.630755263Z       '      }\n' +
    I 2020-12-03T10:51:28.630758361Z       '    ]\n' +
    I 2020-12-03T10:51:28.630761411Z       '  }\n' +
    I 2020-12-03T10:51:28.630764480Z       '}\n',
    I 2020-12-03T10:51:28.630767683Z     headers: {
    I 2020-12-03T10:51:28.630770841Z       'cache-control': 'no-cache, no-store, max-age=0, must-revalidate',
    I 2020-12-03T10:51:28.630774059Z       'content-length': '205',
    I 2020-12-03T10:51:28.630777212Z       'content-type': 'application/json; charset=UTF-8',
    I 2020-12-03T10:51:28.630780501Z       date: 'Thu, 03 Dec 2020 10:51:28 GMT',
    I 2020-12-03T10:51:28.630783676Z       expires: 'Mon, 01 Jan 1990 00:00:00 GMT',
    I 2020-12-03T10:51:28.630786802Z       pragma: 'no-cache',
    I 2020-12-03T10:51:28.630789861Z       server: 'UploadServer',
    I 2020-12-03T10:51:28.630793010Z       vary: 'Origin, X-Origin',
    I 2020-12-03T10:51:28.630796180Z       'x-guploader-uploadid': 'ABg5-UxPaqGQVCgFFluDka8_3PVrUUK73UalrRccAO34uIBZDG-M5TXlcTg7EkY_V5AW7zyZa98hxev6PJmULsXbJILHWadNWw'
    I 2020-12-03T10:51:28.630799671Z     },
    I 2020-12-03T10:51:28.630802693Z     toJSON: [Function: toJSON],
    I 2020-12-03T10:51:28.630812529Z     [Symbol(kCapture)]: false,
    I 2020-12-03T10:51:28.630815801Z     [Symbol(kTransformState)]: {
    I 2020-12-03T10:51:28.630818903Z       afterTransform: [Function: bound afterTransform],
    I 2020-12-03T10:51:28.630822016Z       needTransform: false,
    I 2020-12-03T10:51:28.630825059Z       transforming: false,
    I 2020-12-03T10:51:28.630828106Z       writecb: null,
    I 2020-12-03T10:51:28.630831130Z       writechunk: null,
    I 2020-12-03T10:51:28.630834250Z       writeencoding: 'buffer'
    I 2020-12-03T10:51:28.630837489Z     }
    I 2020-12-03T10:51:28.630844069Z   }
    I 2020-12-03T10:51:28.630847232Z }


Should handle these gracefully with some retries.

--- a/backend/logic/deploy/bucket/gcp.js
+++ b/backend/logic/deploy/bucket/gcp.js
@@ -149,7 +149,7 @@ async function uploadFile(bucket, file, destination, attempt = 0) {
       public: true
     })
   } catch (err) {
-    if (err && err.code === 502 && attempt < 3) {
+    if (err && [502, 503].includes(err.code) && attempt < 3) {
       return await uploadFile(bucket, file, destination, attempt + 1)
     }
 

